zingolib/src/wallet/transaction_records_by_id/input_source.rs

```diff
@@ -0,0 +1,368 @@
+                        zcash_client_backend::wallet::Note::Sapling(note_inner)
+                    })
+                })),
+            ShieldedProtocol::Orchard => Ok(self
+                .get_received_note_from_identifier::<OrchardDomain>(note_record_reference)
+                .map(|note| {
+                    note.map_note(|note_inner| {
+                        zcash_client_backend::wallet::Note::Orchard(note_inner)
+                    })
+                })),
+        }
+    }
+
+    fn select_spendable_notes(
+        &self,
+        account: Self::AccountId,
+        target_value: zcash_primitives::transaction::components::amount::NonNegativeAmount,
+        sources: &[zcash_client_backend::ShieldedProtocol],
+        anchor_height: zcash_primitives::consensus::BlockHeight,
+        exclude: &[Self::NoteRef],
+    ) -> Result<SpendableNotes<ShNoteId>, ZingoLibError> {
+        if account != AccountId::ZERO {
+            return Err(ZingoLibError::Error(
+                "we don't use non-zero accounts (yet?)".to_string(),
+            ));
+        }
+        let mut sapling_note_noteref_pairs: Vec<(sapling_crypto::Note, ShNoteId)> = Vec::new();
+        let mut orchard_note_noteref_pairs: Vec<(orchard::Note, ShNoteId)> = Vec::new();
+        for transaction_record in self.values().filter(|transaction_record| {
+            transaction_record
+                .status
+                .is_confirmed_before_or_at(&anchor_height)
+        }) {
+            if sources.contains(&ShieldedProtocol::Sapling) {
+                sapling_note_noteref_pairs.extend(
+                    transaction_record
+                        .select_unspent_shnotes_and_ids::<SaplingDomain>()
+                        .into_iter()
+                        .filter(|note_ref_pair| !exclude.contains(&note_ref_pair.1)),
+                );
+            }
+            if sources.contains(&ShieldedProtocol::Orchard) {
+                orchard_note_noteref_pairs.extend(
+                    transaction_record
+                        .select_unspent_shnotes_and_ids::<OrchardDomain>()
+                        .into_iter()
+                        .filter(|note_ref_pair| !exclude.contains(&note_ref_pair.1)),
+                );
+            }
+        }
+        let mut sapling_notes = Vec::<ReceivedNote<ShNoteId, sapling_crypto::Note>>::new();
+        let mut orchard_notes = Vec::<ReceivedNote<ShNoteId, orchard::Note>>::new();
+        if let Some(missing_value_after_sapling) = sapling_note_noteref_pairs.into_iter().rev(/*biggest first*/).try_fold(
+            Some(target_value),
+            |rolling_target, (note, noteref)| match rolling_target {
+                Some(targ) => {
+                    sapling_notes.push(
+                        self.get(&noteref.txid).and_then(|tr| tr.get_received_note::<SaplingDomain>(noteref.index))
+                            .ok_or_else(|| ZingoLibError::Error("missing note".to_string()))?
```

Comment:
this error implies we somehow misrepresented the txid across this function. it should never happen, so im fine with this for now